plugin/vcs/internal/oauth/oauth.go

```diff
@@ -0,0 +1,127 @@
+		if err != nil {
+			return nil, errors.Wrapf(err, "construct %s %s", method, url)
+		}
+
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", *token))
+		resp, err := client.Do(req)
+		if err != nil {
+			return nil, errors.Wrapf(err, "%s %s", method, url)
+		}
+		return resp, nil
+	}
+}
+
+// Post makes a HTTP POST request to the given URL using the token. It refreshes
+// token and retries the request in the case of the token has expired.
+func Post(ctx context.Context, client *http.Client, url string, token *string, body io.Reader, tokenRefresher TokenRefresher) (code int, respBody string, err error) {
+	return retry(ctx, client, token, tokenRefresher, requester(ctx, client, http.MethodPost, url, token, body))
+}
+
+// Get makes a HTTP GET request to the given URL using the token. It refreshes
+// token and retries the request in the case of the token has expired.
+func Get(ctx context.Context, client *http.Client, url string, token *string, tokenRefresher TokenRefresher) (code int, respBody string, err error) {
+	return retry(ctx, client, token, tokenRefresher, requester(ctx, client, http.MethodGet, url, token, nil))
+}
+
+// Put makes a HTTP PUT request to the given URL using the token. It refreshes
+// token and retries the request in the case of the token has expired.
+func Put(ctx context.Context, client *http.Client, url string, token *string, body io.Reader, tokenRefresher TokenRefresher) (code int, respBody string, err error) {
+	return retry(ctx, client, token, tokenRefresher, requester(ctx, client, http.MethodPut, url, token, body))
+}
+
+// Delete makes a HTTP DELETE request to the given URL using the token. It refreshes
+// token and retries the request in the case of the token has expired.
+func Delete(ctx context.Context, client *http.Client, url string, token *string, tokenRefresher TokenRefresher) (code int, respBody string, err error) {
+	return retry(ctx, client, token, tokenRefresher, requester(ctx, client, http.MethodDelete, url, token, nil))
+}
+
+const maxRetries = 3
+
+func retry(ctx context.Context, client *http.Client, token *string, tokenRefresher TokenRefresher, f func() (*http.Response, error)) (code int, respBody string, err error) {
+	var resp *http.Response
+	var body []byte
+	for retries := 0; retries < maxRetries; retries++ {
+		select {
+		case <-ctx.Done():
+			return 0, "", ctx.Err()
+		default:
+		}
+
+		resp, err = f()
+		if err != nil {
+			return 0, "", err
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return 0, "", errors.Wrapf(err, "read response body with status code %d", resp.StatusCode)
+		}
+
+		if err = getOAuthErrorDetails(resp.StatusCode, body); err != nil {
+			if _, ok := err.(*oauthError); ok {
+				// Refresh the token
+				if err := tokenRefresher(ctx, client, token); err != nil {
+					return 0, "", err
+				}
+				continue
+			}
+			return 0, "", errors.Errorf("want *oauthError but got %T", err)
```

Comment:
nit: this error log style is more commonly seen in the test. could be improved a bit

Reply:
Which "test" you're referring to? Because it is impossible to get any other type of error back from `getOAuthErrorDetails` in the current implementation, this line of code doesn't look like should be hit at all in practice.

Reply:
I meant "want xxx, but got yyy" usually appears in the test, while it's not that readable if returning to the client.

Maybe just "got unexpected oauth error %T"?

Reply:
> I meant "want xxx, but got yyy" usually appears in the test, while it's not that readable if returning to the client.

Can you link to an example of such "test"? 

> Maybe just "got unexpected oauth error %T"?

I'm OK with this approach

Reply:
I meant the enduser as well as our support folks. The "want xxx, but got yyy" is a bit cryptic to them.

Reply:
Updated in https://github.com/bytebase/bytebase/pull/982/commits/a9570518b65aef4a0f187d2df158c3e18aedfcfe.